File: wordle.pl
#!/usr/bin/env perl
# https://www.nytimes.com/games/wordle/
use strict;
use warnings;
use Data::Dumper;
use feature 'say';

my @dict = sub {
    my %hash;
    open my $dict, '<', shift || die;
    while(<$dict>){
        chomp;
        $hash{uc($_)}++ if length == 5;
    }
    keys(%hash);
}->('/usr/share/dict/words');

my $apperance = [];
for my $nth (0..3){
    $apperance->[$nth] = {};
    for my $char ('A'..'Z'){
        $apperance->[$nth]->{$char} = {};
        my $regex = '^' . '.' x $nth . $char;
        my @conn_words = grep(/$regex/, @dict);
        $regex = '.' x ($nth+1)  . '([A-Z])';
        my @next_chars = map(/$regex/, @conn_words);
        map($apperance->[$nth]->{$char}->{$_}++, @next_chars);
    }
}

say Dumper $apperance;

@dict = grep(/A/, @dict);
@dict = grep(/^A/, @dict);
@dict = grep(/R/, @dict);
@dict = grep(/^.R/, @dict);
@dict = grep(/I/, @dict);
@dict = grep(/^..I/, @dict);
@dict = grep(/S/, @dict);
@dict = grep(/^...S/, @dict);
@dict = grep(/E/, @dict);
@dict = grep(/^....E/, @dict);

map {say} (sort @dict);

my ($hash, $order) = ({}, {});

for my $position (0..4){
    $hash->{$position} = {};
    $order->{$position} = [];
    for my $word (@dict){
        $hash->{$position}->{substr($word, $position, 1)}++;
    }
    for my $char (sort {
        $hash->{$position}->{$b} - $hash->{$position}->{$a}
                  } keys(%{$hash->{$position}})){
        push @{$order->{$position}}, $char;
    }
}

while(1){
    my @buf;
    for my $position (0..4){
        if(my $char = shift @{$order->{$position}}){
            push @buf, sprintf('%4d %s', $hash->{$position}->{$char}, $char);
        }else{
            push @buf, '      ';
        }
    }

    my $s = join('  ', @buf);
    exit if $s =~ /^\s+$/;
    say $s;
}

exit;

File: wordle.rb
#!/usr/bin/env ruby
# https://www.nytimes.com/games/wordle/

Gray = 'gray'
Yellow = 'yellow'
Green = 'green'

class Wordle
  def hint(*hint)
    position = 0
    hint.each do |hash|
      position += 1
      char = hash.keys[0]
      case hash[char]
      when Gray then
        duplicated_gray = false
        # 同じ文字でグリーンが存在する場合、
        # グレーは「存在しない」ではなく「重複しない」という意味になる
        hint.each do |hash|
          if char === hash.keys[0] then
            if hash[hash.keys[0]] === Green then
              duplicated_gray = true
            end
          end
        end
        if duplicated_gray then
          @dict = @dict.grep_v(/#{char}.*#{char}/)
        else
          @dict = @dict.grep_v(/#{char}/)
        end
      when Yellow then
        @dict = @dict.grep(/#{char}/).grep_v(/^#{'.' * (position-1)}#{char}/)
      when Green then
        @dict = @dict.grep(/^#{'.' * (position-1)}#{char}/)
      end
    end
  end

  def initialize
    @dict = []
    File.foreach('/usr/share/dict/words') do |line|
      line.chomp!
      if line.length == 5 then
        @dict.push(line.upcase)
      end
    end
  end

  def report
    @dict.each do |word|
      puts word
    end

    # 何文字めに、どの文字が何回出現しているかを格納
    occurrences = {}
    order = []

    for position in 0..4 do
      occurrences[position] = {}
      @dict.each do |word|
        char = word.slice(position)
        if occurrences[position][char] == nil then
          occurrences[position][char] = 1
        else
          occurrences[position][char] += 1
        end
      end

      order[position] = []
      occurrences[position].keys.sort do |a, b|
        occurrences[position][b] <=> occurrences[position][a]
      end.each do |char|
        order[position].push(char);
      end
    end

    loop do
      buf = []
      for position in 0..4 do
        if char = order[position].shift then
          buf.push(sprintf('%4d %s', occurrences[position][char], char))
        else
          buf.push('      ')
        end
      end

      s = buf.join('  ')
      if s.match(/^\s+$/) then
        exit
      end
      puts s
    end
  end
end

w = Wordle.new
w.hint({A: Gray},
       {R: Gray},
       {I: Yellow},
       {S: Green},
       {E: Gray})
w.hint({K: Gray},
       {I: Green},
       {S: Gray},
       {S: Green},
       {Y: Gray})
w.report
